tests/e2e/refcount_test.go

```diff
@@ -0,0 +1,150 @@
+
+// Test for refcount recovery
+// * Create a volume
+// * Attach 5 containers to it with restart flag
+// * Verify attached status after each attach
+// * Check the refcount log in log file
+// * Verify the content on volume created by containers
+// * Check the volume entry on mount file : /proc/mounts
+// * Try to delete volume. Should Fail
+// * Kill Docker (ungraceful)
+// * Check the recovery log record in the log file
+// * Cleanup the containers
+// * Delete the volume. Should succeed
+// * There should be no entry in /proc/mounts
+
+func (s *RefcountTestSuite) TestRefcountingRecovery(c *C) {
+	misc.LogTestStart(c.TestName())
+
+	// Check if volume is available in docker host
+	accessible := verification.CheckVolumeAvailability(s.config.DockerHosts[0], s.volumeName)
+	c.Assert(accessible, Equals, true, Commentf("Volume %s is not available", s.volumeName))
+
+	numContainers := 5
+
+	// Each container creates a file with its index as the filename suffix. After that it keeps running.
+	for i := 0; i < numContainers; i++ {
+		cName := inputparams.GetUniqueContainerName(c.TestName())
+		out, err := dockercli.AttachWriteVolumeWithRestart(s.config.DockerHosts[0], s.volumeName, cName, "file_"+strconv.Itoa(i))
+		c.Assert(err, IsNil, Commentf(out))
+
+		s.containerNameList = append(s.containerNameList, cName)
+
+		status := verification.VerifyAttachedStatus(s.volumeName, s.config.DockerHosts[0], s.config.EsxHost)
+		c.Assert(status, Equals, true, Commentf("Volume %s is not attached", s.volumeName))
+	}
+
+	// The refcount log should indicate the refcount of volume. It should be equal to number of containers
+	isRefcounted := verification.CheckRefCountInLogs(s.config.DockerHosts[0], s.volumeName, 1, 5)
+	c.Assert(isRefcounted, Equals, true, Commentf("Volume %s refcount error", s.volumeName))
+
+	// Verify all the files created can be seen from a different container
+	listFiles, err := dockercli.ListFilesOnVolume(s.config.DockerHosts[0], s.volumeName)
+	for i := 0; i < numContainers; i++ {
+		c.Assert(strings.Contains(listFiles, "file_"+strconv.Itoa(i)), Equals, true,
+			Commentf("File %d is not available in %s", listFiles))
+	}
+
+	// Verify the entry of volume in mount file
+	isAvailable := verification.CheckVolumeInProcMounts(s.config.DockerHosts[0], s.volumeName)
+	c.Assert(isAvailable, Equals, true, Commentf("Volume %s unavailable in /proc/mounts", s.volumeName))
+
+	// This delete should fail since the volume is still being used
+	out, err := dockercli.DeleteVolume(s.config.DockerHosts[0], s.volumeName)
+	c.Assert(err, Not(IsNil), Commentf(out))
+
+	// Kill Docker (ungraceful). It restarts automatically
+	out, err = dockercli.KillDocker(s.config.DockerHosts[0])
+	c.Assert(err, IsNil, Commentf(out))
+
+	// Verify the recovery record. There is a delay after which recovery record appears in the log
+	// due to refcounting being delayed in background. (Docker doesn't respond immediately)
+	for i := 0; i < 60; i++ {
```

Comment:
defined const under a separate util which polls the method to check recovery record.